src/equations/equations.jl

```diff
@@ -220,6 +238,12 @@ example of equations with nonconservative terms.
 The return value will be `True()` or `False()` to allow dispatching on the return type.
 """
 have_nonconservative_terms(::AbstractEquations) = False()
+"""
+    nnoncons(equations)
+Number of nonconservative terms for a particular equation. The default is 0 and 
+it must be defined for each nonconservative equation independently.
+"""
+nnoncons(::AbstractEquations) = 0
```

Comment:
```suggestion

"""
    nnoncons(equations)
Number of nonconservative terms for a particular equation. The default is 0 and 
it must be defined for each nonconservative equation independently.
"""
nnoncons(::AbstractEquations) = 0

```
Also, could we make this name slightly more descriptive, e.g., `n_nonconservative_terms` (consistent with `have_nonconservative_terms`)?

Reply:
Update: It seems to me that this function is not used at all in the implementation and could be removed - or did I miss it?

Reply:
> Update: It seems to me that this function is not used at all in the implementation and could be removed - or did I miss it?

I see that many functions like this are defined in equations.jl with an explanatory docstring, e.g., `have_nonconservative_terms`, `have_constant_speed`, `cons2prim`, `prim2cons`, `entropy`. Since the functions used by the solver are actually the ones inside the equations, I assumed that the reason for this was to provide a single docstring. Is it then better to define this variable with the docstring in ideal_glm_mhd_2d.jl and then copy the docstring to all future implementations in non-conservative equations? (e.g., ideal_glm_mhd_1d.jl, ideal_mhd_multicomponent_*d.jl, the shallow-water equations, in the future ideal_mhd_multiion_*d.jl, etc.)

Reply:
> Also, could we make this name slightly more descriptive, e.g., n_nonconservative_terms (consistent with have_nonconservative_terms)?

Makes sense

Reply:
> > Update: It seems to me that this function is not used at all in the implementation and could be removed - or did I miss it?
> 
> I see that many functions like this are defined in equations.jl with an explanatory docstring, e.g., `have_nonconservative_terms`, `have_constant_speed`, `cons2prim`, `prim2cons`, `entropy`. Since the functions used by the solver are actually the ones inside the equations, I assumed that the reason for this was to provide a single docstring. Is it then better to define this function with the docstring in ideal_glm_mhd_2d.jl and then copy the docstring to all future implementations in non-conservative equations? (e.g., ideal_glm_mhd_1d.jl, ideal_mhd_multicomponent_*d.jl, the shallow-water equations, in the future ideal_mhd_multiion_*d.jl, etc.)

No, it's just that I didn't find a place where `nnoncons(...)` was used in the code anywhere - I wonder if it can just be remove entirely

Reply:
It is used in dg_2d_subcell_limiters.jl. E.g.,
https://github.com/trixi-framework/Trixi.jl/blob/c7c3ca0fb17bddce532589eaae422fa061fe86f0/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L256
and
https://github.com/trixi-framework/Trixi.jl/blob/c7c3ca0fb17bddce532589eaae422fa061fe86f0/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L301
and of course in the allocation of the multi-dimensional array `fhat_noncons_temp`.

Reply:
Ah, sorry, I missed that. In that case, I propose to just rename it but keep it organized as it is right now.

However, the question is whether you want to provide a default implementation that equals to zero. Wouldn't it be better to have no implementation at all and require all non-cons systems to implement their own version?

Otherwise you could run, e.g., the SWE with this limiter and it would silently assume that there are no non-cons variables. As long as n_nonconservative_terms is never called outside a scope where `have_nonconservative_terms` is true, this shouldn't affect all conservative systems and hopefully reduce accidental wrong usage.

Reply:
> Wouldn't it be better to have no implementation at all and require all non-cons systems to implement their own version?

Yeah, that makes sense!
I renamed the function and removed the default output in 79109937d.

Reply:
After looking at the CI tests, I realized that `n_nonconservative_terms` is actually used for all equations while creating the cache in
https://github.com/amrueda/Trixi.jl/blob/3f9f0fec87e465ba0a3e98ec6ac16600bd18bc8e/src/solvers/dgsem_tree/dg_2d_subcell_limiters.jl#L29-L32

We can think about only adding `flux_nonconservative_temp_threaded`, `fhat_nonconservative_temp_threaded` and `phi_threaded` to the cache, if `have_nonconservative_terms()==true`.

Reply:
I was faster:rocket::smile:
What I meant above was something like that:
```
if have_nonconservative_terms(equations)
    flux_nonconservative_temp_threaded = A4d[A4d(undef, nvariables(equations),
                                                 n_nonconservative_terms(equations),
                                                 nnodes(dg), nnodes(dg))
                                             for _ in 1:Threads.nthreads()]
    fhat_nonconservative_temp_threaded = A4d[A4d(undef, nvariables(equations),
                                                 n_nonconservative_terms(equations),
                                                 nnodes(dg), nnodes(dg))
                                             for _ in 1:Threads.nthreads()]

    phi_threaded = A4d[A4d(undef, nvariables(equations),
                           n_nonconservative_terms(equations),
                           nnodes(dg), nnodes(dg))
                       for _ in 1:Threads.nthreads()]
    cache = (; cache..., flux_nonconservative_temp_threaded, fhat_nonconservative_temp_threaded, phi_threaded)
end
```

Reply:
Thanks! I'll implement something like that.

Reply:
Done in 0eee72182.

Reply:
@ranocha does this still result in a type stable cache? I'd say yes, since the result depends on `have_nonconservative_terms(equations)` is a compile-time constant, and `create_cache` itself will specialize on `equations`.

Otherwise, @amrueda @bennibolm what you could do is to call `create_cache_nonconservative(..., have_nonconservative_terms(equations))` at the end of your `create_cache`, and make it return an empty named tuple for all conservative models.